automerge-backend/src/event_handlers.rs

```diff
@@ -0,0 +1,67 @@
+use std::fmt::Debug;
+
+use crate::Change;
+
+#[derive(Clone, Copy)]
+pub struct EventHandlerId(usize);
+
+/// A sequence of event handlers.
+///
+/// This maintains the order of insertion so handlers will be called in a consistent order.
+#[derive(Debug, Default)]
+pub struct EventHandlers(Vec<EventHandler>);
+
+impl Clone for EventHandlers {
+    fn clone(&self) -> Self {
+        EventHandlers(Vec::new())
+    }
+}
+
+impl EventHandlers {
+    pub(crate) fn before_apply_change(&mut self, change: &Change) {
+        for handler in &mut self.0 {
+            if let EventHandler::BeforeApplyChange(f) = handler {
+                f.0(change)
+            }
+        }
+    }
+
+    pub(crate) fn after_apply_change(&mut self, change: &Change) {
+        for handler in &mut self.0 {
+            if let EventHandler::AfterApplyChange(f) = handler {
+                f.0(change)
+            }
+        }
+    }
+
+    /// Remove the handler with the given id.
+    pub fn remove_handler(&mut self, id: EventHandlerId) {
+        self.0.remove(id.0);
```

Comment:
This can panic if the index is out of bounds. The `EventHandlerId` wrapper makes this less likely to happen but you could somehow get an ID from a different document.

Reply:
Ah thanks, good catch. I'll fix